Feature/a12n

HerdaSinama opened the pull request:
hello everyone its my 1st pr! I'll put [it](https://gist.github.com/zmts/802dc9c3510d79fd40f9dc38a12bccfc) here so I don't forget

HerdaSinama committed:
Init a12n app and set settings

HerdaSinama committed:
Add ser. and view for registration

- UserSerializer: have validate method for password and create user (field password and method set_password must be define in User model)
- RegistrationAPIView: 1st try jwt

HerdaSinama committed:
Add model UserSession and ser. UserLoginSerializer

- UserSession: model for storing refresh token sessions, which will include information about the user and the token
- UserLoginSerializer: serializer checks the validity of the data. if data is valid, check the count of sessions; create a new session; create a session in the db

HerdaSinama committed:
Fix serializer UserLoginSerializer

Fix:
- orm error (get together first)
- error deleting all sessions instead of the last one
- return error (invalid data)

HerdaSinama committed:
Add view LoginAPIView

HerdaSinama committed:
Fix model UserSession field and type

- now refresh_token type: CharField
- expires_at rename to expires_in

HerdaSinama committed:
Add view LogoutAPIView

HerdaSinama committed:
Add in ser. UserRegisterSerializer validate

add:
- checking for uniqueness userbane and email;
- init additional checks on the correctness of the registration data

HerdaSinama committed:
Fix ser. UserRegisterSerializer

fix:
- the processing cycle for each username and email verification condition has been removed. now return general error message.
- def validate ORM request fixed

HerdaSinama committed:
Add view RegistrationAPIView

HerdaSinama committed:
Add view ActivateAccountAPIView

HerdaSinama committed:
Fix view LogoutAPIView

- added more information in Response

HerdaSinama committed:
Doc view

RegistrationAPIView;
ActivateAccountAPIView;
LoginAPIView;
LogoutAPIView;

HerdaSinama commented:
forgot to add y -_-

ArtyomYapryntsev commented on server/apps/a12n/models.py:
```diff
@@ -0,0 +1,18 @@
+from django.db import models
+from apps.acсount.models import User
```
In this pull request context the `account` apps does not exist.
So, u can not to import the `User` model.

Also, the django recommendation to use the `User` model is:
- using the `django.contrib.auth.get_user_model()`
- using the `django.conf.settings.AUTH_USER_MODEL`

U can read this [resource](https://docs.djangoproject.com/en/5.1/topics/auth/customizing/#referencing-the-user-model) for more information.

ArtyomYapryntsev commented on server/apps/a12n/models.py:
```diff
@@ -0,0 +1,18 @@
+        User,
+        on_delete=models.CASCADE,
+    )
+    refresh_token = models.CharField(
```
Usually, "refresh tokens" are not used in the context of session authentication.
It may be better to rename this field to "token".

What do you think about it?

HerdaSinama replied:
i try to made token auth) I used [this](https://gist.github.com/zmts/802dc9c3510d79fd40f9dc38a12bccfc#%D0%BB%D0%BE%D0%B3%D0%B8%D0%BD-%D1%81%D0%BE%D0%B7%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5-%D1%81%D0%B5%D1%81%D1%81%D0%B8%D0%B8%D1%82%D0%BE%D0%BA%D0%B5%D0%BD%D0%BE%D0%B2-apiauthlogin) guide (it seemed to me the most logical)

ArtyomYapryntsev replied:
Ok, I get it, then keep it up!!!

ArtyomYapryntsev commented on server/apps/a12n/models.py:
```diff
@@ -0,0 +1,18 @@
+    created_at = models.DateTimeField(auto_now_add=True,)
+
+    def __str__(self):
+        return f'{self.user.username} - {self.refresh_token}'
```
In this case, to print convert the `UserSession` instance u need to select the related user info for the `user.username` field using. So, if u want to convert the list of `UserSession` to the list og strings u need to select the related user for an each instance.
Perhaps if you use `self.user_id` instead of `self.user.username`, it will also be clear?

ArtyomYapryntsev commented on server/apps/a12n/serializers.py:
```diff
@@ -0,0 +1,112 @@
+            'password_confirm',
+        ]
+
+    def validate_username(self, value):
+        pass
+
+    def validate_email(self, value):
+        pass
+
+    def validate_first_name(self, value):
+        pass
+
+    def validate_last_name(self, value):
+        pass
+
+    def validate_password(self, value):
+        pass
+
```
Why are these methods needed here?

HerdaSinama replied:
validation so that the fields don't look like (username: "-=/\|; firstname: serega228~ etc) может такое уже есть мне в падлу про это читать - удалил.

ArtyomYapryntsev commented on server/apps/a12n/serializers.py:
```diff
@@ -0,0 +1,112 @@
+        
+        user_session = UserSession.objects.filter(user=user)
+        if user_session.count() > 5:
+            old_session = user_session.order_by('created_at').first()
+            if old_session:
+                old_session.delete()
```
I do not check it, but maybe exists some way to delete the first instance with only one query?

Smth lik: `user_session.order_by(...).limit(1).delete()`?

ArtyomYapryntsev commented on server/apps/a12n/serializers.py:
```diff
@@ -0,0 +1,112 @@
+
+import re
+
+from apps.acсount.models import User
```
Same, readd [this](https://docs.djangoproject.com/en/5.1/topics/auth/customizing/#referencing-the-user-model).

ArtyomYapryntsev commented on server/apps/a12n/views.py:
```diff
@@ -0,0 +1,163 @@
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework_simplejwt.tokens import RefreshToken
+
+from .serializers import UserRegisterSerializer, UserLoginSerializer
+from .models import UserSession
+from apps.account.models import User
+
+from django.contrib.auth.tokens import default_token_generator
+from django.utils.http import urlsafe_base64_decode, urlsafe_base64_encode
+from django.utils.encoding import force_bytes, force_str
+from django.core.mail import BadHeaderError, send_mail
+
```
Good practice to correctly sort your imports (read some resources).
U can use the `isort` tool that can refactor your import autormically or just check your code on errors.

ArtyomYapryntsev commented on server/apps/a12n/views.py:
```diff
@@ -0,0 +1,163 @@
+            confirm_url = f'/confirm/{uid}/{token}'
+            
+            try:
+                send_mail(
+                    'Register confirm',
+                    f'Hi, {user.username}, please confirm your registration,
+                    follow this link: {confirm_url}',
+                    'example@example.com',
+                    [user.email],
+                )
```
Good job :smiley:

ArtyomYapryntsev commented on server/apps/a12n/views.py:
```diff
@@ -0,0 +1,163 @@
+                    status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                )
+            
+        return Response(
+            serializer.errors,
+            status=status.HTTP_400_BAD_REQUEST,
+            )
```
If u will test, u will receive a 400 error regardless of sending the email.
Try to solve it yourself.

ArtyomYapryntsev commented on server/apps/a12n/views.py:
```diff
@@ -0,0 +1,163 @@
+        if serializer.is_valid():
+            data = serializer.validated_data
+
+            response = Response(
+                {'access_token': data['access_token']},
+                status=status.HTTP_200_OK
+            )
```
Why u need to build the response here?

HerdaSinama replied:
idk

ArtyomYapryntsev reviewed (changes requested):
Good work, the implementation of registration via email and account activation via separate links - good training :+1:.

I think you're confusing the concepts of the session and token authentications (we can discuss about that later). Also read about how to reference a custom model.

HerdaSinama committed:
Fix view

- sort import
- import get_user_model
- RegistrationAPIView method post add Response (create account 201)

HerdaSinama committed:
Fix ser.

- now if count sessions > 5 deleted all sessions

HerdaSinama committed:
Add ser. GenerateNewTokenSerializer

HerdaSinama committed:
Fix ser.

- UserLoginSerializer: method validate split on validate and create
- GenerateNewTokenSerializer: optimized database query (method save)

HerdaSinama committed:
Add doc for ser.

HerdaSinama committed:
Add view TokensRefreshAPIView
